pkg/graveler/ref/tag_iterator.go

```diff
@@ -2,118 +2,77 @@ package ref
-	value        *graveler.TagRecord
-	buf          []*graveler.TagRecord
-	offset       string
-	fetchSize    int
-	err          error
-	state        iteratorState
+type KVTagIterator struct {
+	it     kv.MessageIterator
+	err    error
+	value  *graveler.TagRecord
+	repoID string
+	store  kv.Store
 }
 
-type tagRecord struct {
-	TagID    graveler.TagID    `db:"id"`
-	CommitID graveler.CommitID `db:"commit_id"`
-}
-
-func NewTagIterator(ctx context.Context, db db.Database, repositoryID graveler.RepositoryID, fetchSize int) *TagIterator {
-	return &TagIterator{
-		db:           db,
-		ctx:          ctx,
-		repositoryID: repositoryID,
-		fetchSize:    fetchSize,
-		buf:          make([]*graveler.TagRecord, 0, fetchSize),
+func NewKVTagIterator(ctx context.Context, store *kv.StoreMessage, repositoryID graveler.RepositoryID) (*KVTagIterator, error) {
+	it, err := kv.NewPrimaryIterator(ctx, store.Store, (&graveler.TagData{}).ProtoReflect().Type(),
+		graveler.TagPartition(repositoryID.String()),
+		graveler.TagPath(""), "")
+	if err != nil {
+		return nil, err
 	}
+	return &KVTagIterator{
+		it:     it,
+		store:  store.Store,
+		repoID: repositoryID.String(),
+	}, nil
 }
 
-func (ri *TagIterator) Next() bool {
-	if ri.err != nil {
+func (i *KVTagIterator) Next() bool {
+	if i.Err() != nil {
 		return false
 	}
-
-	ri.maybeFetch()
-
-	// stage a value and increment offset
-	if len(ri.buf) == 0 {
+	if !i.it.Next() {
+		i.value = nil
 		return false
 	}
-	ri.value = ri.buf[0]
-	ri.buf = ri.buf[1:]
-	ri.offset = string(ri.value.TagID)
-	return true
-}
-
-func (ri *TagIterator) maybeFetch() {
-	if ri.state == iteratorStateDone {
-		return
-	}
-	if len(ri.buf) > 0 {
-		return
-	}
-
-	var offsetCondition string
-	if ri.state == iteratorStateInit {
-		offsetCondition = iteratorOffsetCondition(true)
-		ri.state = iteratorStateQuerying
-	} else {
-		offsetCondition = iteratorOffsetCondition(false)
-	}
-
-	var buf []*tagRecord
-	err := ri.db.Select(ri.ctx, &buf, `
-			SELECT id, commit_id
-			FROM graveler_tags
-			WHERE repository_id = $1
-			AND id `+offsetCondition+` $2
-			ORDER BY id ASC
-			LIMIT $3`, ri.repositoryID, ri.offset, ri.fetchSize)
-	if err != nil {
-		ri.err = err
-		return
-	}
-	if len(buf) < ri.fetchSize {
-		ri.state = iteratorStateDone
+	e := i.it.Entry()
+	if e == nil {
+		i.err = graveler.ErrNotFound
+		return false
 	}
-	for _, b := range buf {
-		ri.buf = append(ri.buf, &graveler.TagRecord{
-			TagID:    b.TagID,
-			CommitID: b.CommitID,
-		})
+	tag := e.Value.(*graveler.TagData)
+	i.value = &graveler.TagRecord{
+		TagID:    graveler.TagID(tag.Id),
+		CommitID: graveler.CommitID(tag.CommitId),
 	}
+	return true
 }
 
-func (ri *TagIterator) SeekGE(id graveler.TagID) {
-	if errors.Is(ri.err, ErrIteratorClosed) {
-		return
-	}
-	ri.offset = string(id)
-	ri.buf = ri.buf[:0]
-	ri.value = nil
-	ri.err = nil
-	ri.state = iteratorStateInit
+func (i *KVTagIterator) SeekGE(id graveler.TagID) {
+	ctx := context.Background()
```

Comment:
ctx should be part of the iterator struct and passed in 'New'

Reply:
In that case, the tag iterator implements value iterator and more iterators implement it such as commit, and branch.
If we will add it as part of the tag iterator we will need to do it in all of them, and I don't think it's necessary at this point. WDYT?

Reply:
Constructing iterators isn't part of the interface, so why passing it in `NewKVTagIterator` will require a change to other iterators?

Reply:
See example of SecondaryIterator

Reply:
done